src/distribution/cauchy.rs

```diff
@@ -0,0 +1,232 @@
+
+/// A Cauchy distribution.
+///
+/// A Cauchy distribution (aka Lorentz or Cauchy–Lorentz distribution) is a continuous
+/// probability distribution with a location parameter `x_0`, a scale parameter `gamma > 0`,
+/// and the following probability density function:
+///
+/// `p(x) = const / (1 + ((x - x_0) / gamma)^2)`.
+///
+/// A Cauchy distribution is long tailed and has no well-defined mean or variance. It is
+/// unimodal with its mode at `x_0`, around which it is symmetric. The ratio of two
+/// independent Gaussian distributed random variables is Cauchy distributed.
+///
+/// See [Wikipedia article on Cauchy
+/// distribution](https://en.wikipedia.org/wiki/Cauchy_distribution).
+#[derive(Clone, Copy, Debug)]
+pub struct Cauchy {
+    x_0: f64,
+    gamma: f64,
+}
+
+impl Cauchy {
+    /// Create a Cauchy distribution with location `x_0` and scale `gamma`.
+    ///
+    /// It should hold that `gamma > 0`.
+    #[inline]
+    pub fn new(x_0: f64, gamma: f64) -> Self {
+        should!(gamma > 0.0);
+        Cauchy { x_0, gamma }
+    }
+
+    /// Return the location parameter.
+    #[inline(always)]
+    pub fn x_0(&self) -> f64 {
+        self.x_0
+    }
+
+    /// Return the scale parameter.
+    #[inline(always)]
+    pub fn gamma(&self) -> f64 {
+        self.gamma
+    }
+}
+
+impl distribution::Continuous for Cauchy {
+    #[inline]
+    fn density(&self, x: f64) -> f64 {
+        use std::f64::consts::PI;
+        let deviation = x - self.x_0;
+        self.gamma / (PI * (self.gamma * self.gamma + deviation * deviation))
+    }
+}
+
+impl distribution::Distribution for Cauchy {
+    type Value = f64;
+
+    #[inline]
+    fn distribution(&self, x: f64) -> f64 {
+        use std::f64::consts::FRAC_1_PI;
+        FRAC_1_PI * ((x - self.x_0) / self.gamma).atan() + 0.5
+    }
+}
+
+impl distribution::Entropy for Cauchy {
+    #[inline]
+    fn entropy(&self) -> f64 {
+        (std::f64::consts::PI * 4.0 * self.gamma).ln()
+    }
+}
+
+impl distribution::Inverse for Cauchy {
+    #[inline]
+    fn inverse(&self, p: f64) -> f64 {
+        use std::f64::{consts::PI, INFINITY, NEG_INFINITY};
+
+        should!((0.0..=1.0).contains(&p));
+
+        if p <= 0.0 {
+            NEG_INFINITY
+        } else if 1.0 <= p {
+            INFINITY
+        } else {
+            self.x_0 + self.gamma * (PI * (p - 0.5)).tan()
+        }
+    }
+}
+
+impl distribution::Median for Cauchy {
+    #[inline]
+    fn median(&self) -> f64 {
+        self.x_0
+    }
+}
+
+impl distribution::Modes for Cauchy {
+    #[inline]
+    fn modes(&self) -> Vec<f64> {
+        vec![self.x_0]
+    }
+}
+
+impl distribution::Sample for Cauchy {
+    #[inline]
+    fn sample<S>(&self, source: &mut S) -> f64
+    where
+        S: Source,
+    {
+        let gaussian = distribution::Gaussian::new(0.0, 1.0);
+        let a = gaussian.sample(source);
+        let b = gaussian.sample(source);
+        self.x_0() + self.gamma() * a / (b.abs() + f64::EPSILON)
```

Comment:
Were you not planning to update this? There is also one in `sample_indirectly`. In general, is this the right way of handing this? Perhaps one could have a loop to generate a non-zero `b`.

Reply:
Ah yes, I must have forgotten to `git add`. Fixed it.

Adding a tiny value to a nonnegative enumerator to avoid division by zero is a common approach in fields I'm familiar with (physics and machine learning), but I'm not an expert in numerical libraries. If you're uncomfortable with it then I'd rather suggest replacing the sampling algorithm with the more naive method (i.e., `self.inverse(uniform random sample)`) since it has constant runtime and is easier to understand.

I would have avoided a regularizer like this if the denominator was supplied by the user. But since the denominator (`b.abs()`) is a random sample, the probability that adding `f64::MIN_POSITIV` actually has any effect is very close to (if not numerically exactly) zero, and in the cases where it does have an effect, the result has low precision anyway.